[ARCHETYPE-406][ARCHETYPE-558][ARCHETYPE-562] Support full Velocity default value expressions for required properties; properly handle ordering among same

mbenson opened the pull request:
* [ARCHETYPE-406]
* [ARCHETYPE-558]
* [ARCHETYPE-562]

Support full Velocity default value expressions for required properties; properly handle ordering among same

mbenson commented:
Also supercedes https://github.com/apache/maven-archetype/pull/61

elharo commented on maven-archetype-plugin/src/main/java/org/apache/maven/archetype/ui/generation/DefaultArchetypeGenerationConfigurator.java:
```diff
@@ -159,6 +169,11 @@ else if ( archetypeArtifactManager.isOldArchetype( ad.getGroupId(), ad.getArtifa
         }
 
+        List<String> propertiesRequired = archetypeConfiguration.getRequiredProperties();
+        getLogger().debug( "Required properties before content sort: " + propertiesRequired );
```
please take these logging statements out before committing.

elharo commented on maven-archetype-plugin/src/main/java/org/apache/maven/archetype/ui/generation/DefaultArchetypeGenerationConfigurator.java:
```diff
@@ -168,65 +183,54 @@ else if ( archetypeArtifactManager.isOldArchetype( ad.getGroupId(), ad.getArtifa
 
-                                value = getTransitiveDefaultValue( value, archetypeConfiguration, requiredProperty,
-                                                                   context );
+                                value = getTransitiveDefaultValue( value, requiredProperty, context );
```
it would be much clearer not to reuse the local variables

elharo commented on maven-archetype-plugin/src/main/java/org/apache/maven/archetype/ui/generation/DefaultArchetypeGenerationConfigurator.java:
```diff
@@ -313,46 +320,22 @@ else if ( !archetypeGenerationQueryer.confirmConfiguration( archetypeConfigurati
+                Velocity.evaluate( context, stringWriter, requiredProperty, defaultValue );
+                return stringWriter.toString();
+            }
+            catch ( Exception ex )
```
can this exception be made more specific?

elharo commented on maven-archetype-plugin/src/main/java/org/apache/maven/archetype/ui/generation/DefaultArchetypeGenerationConfigurator.java:
```diff
@@ -368,71 +351,113 @@ private void restoreCommandLineProperties( ArchetypeConfiguration archetypeConfi
+                            }
+                        }, referencedPropertyNames );
+                    }
+                    catch ( ParseException e )
```
if an empty catch block is really OK here, please add a comment explaining why

elharo commented on maven-archetype-plugin/src/main/java/org/apache/maven/archetype/ui/generation/DefaultArchetypeGenerationConfigurator.java:
```diff
@@ -368,71 +351,113 @@ private void restoreCommandLineProperties( ArchetypeConfiguration archetypeConfi
+        }
+
+        /**
+         * Yoda semantics.
```
I don't understand this comment. Please rephrase.

elharo commented on maven-archetype-plugin/src/main/java/org/apache/maven/archetype/ui/generation/DefaultArchetypeGenerationConfigurator.java:
```diff
@@ -368,71 +351,113 @@ private void restoreCommandLineProperties( ArchetypeConfiguration archetypeConfi
+                }
             }
-            if ( "package".equals( left ) )
+            Set<String> references = propertyReferenceMap.get( sourceProperty );
```
It's mildly confusing to have a local variable and method with the same name. Can you rename one?

mbenson commented:
Hi @elharo ... incidentally in my semester of Java your book was my textbook... which ultimately led me to a 20-year-plus-and-running career, so thanks!
Several of your notes were made against code that existed before I started making changes. I don't mind making the changes, but could you confirm that you prefer to see the changes made as part of this request? In the meantime I'll work on addressing the comments that apply to my own work.

elharo reviewed (commented):
If you've moved a line, it counts as new code and should be brought up to spec; but don't touch other lines you would not otherwise touch.

mbenson commented:
I believe I have addressed all requested modifications, but note that the ML received no notification of same. Thanks for your consideration of these changes.

elharo reviewed (commented):
checking jenkins: https://ci-builds.apache.org/job/Maven/job/maven-box/job/maven-archetype/job/ARCHETYPE-406/

mbenson commented:
Looks like that succeeded, yes?

elharo commented on maven-archetype-plugin/src/main/java/org/apache/maven/archetype/ui/generation/DefaultArchetypeGenerationConfigurator.java:
```diff
@@ -313,46 +311,23 @@ else if ( !archetypeGenerationQueryer.confirmConfiguration( archetypeConfigurati
-
-                result = StringUtils.replace( result, "${" + property + "}",
-                                              archetypeConfiguration.getProperty( property ) );
+                Velocity.evaluate( context, stringWriter, requiredProperty, defaultValue );
```
is defaultValue correctly named? This looks like a template, not a default value, but I'm not a velocity expert.

mbenson replied:
This is the default value of some required property of the archetype, which in this context has been determined to appear to contain expansion syntax (`${`) indicative of the whole being intended to be interpreted as a Velocity expression. Does that make sense?

elharo replied:
Not, not really. That's where it came from, but it's not what it is inside this method or inside Velocity.evaluate. It's a template to be interpreted.

elharo commented on maven-archetype-plugin/src/main/java/org/apache/maven/archetype/ui/generation/DefaultArchetypeGenerationConfigurator.java:
```diff
@@ -313,46 +311,23 @@ else if ( !archetypeGenerationQueryer.confirmConfiguration( archetypeConfigurati
+            }
+            catch ( IOException ex )
+            {
+                // closing StringWriter shouldn't actually generate any exception
```
Yes, but wrap in a runtime exception just in case.

elharo commented on maven-archetype-plugin/src/main/java/org/apache/maven/archetype/ui/generation/DefaultArchetypeGenerationConfigurator.java:
```diff
@@ -313,46 +311,25 @@ else if ( !archetypeGenerationQueryer.confirmConfiguration( archetypeConfigurati
+            {
+                // closing StringWriter shouldn't actually generate any exception
+                throw new RuntimeException(
+                                String.format("Exception closing %s", StringWriter.class.getSimpleName()), ex);
```
no need to use String.format or getClass here. This is always a constant string.

elharo commented on maven-archetype-plugin/src/main/java/org/apache/maven/archetype/ui/generation/DefaultArchetypeGenerationConfigurator.java:
```diff
@@ -313,46 +311,25 @@ else if ( !archetypeGenerationQueryer.confirmConfiguration( archetypeConfigurati
 
-    private String getTransitiveDefaultValue( String defaultValue, ArchetypeConfiguration archetypeConfiguration,
-                                              String requiredProperty, Context context )
+    private String getTransitiveDefaultValue( String defaultValue, String requiredProperty, Context context )
```
This method name is unclear. `evaluateTemplate` would be clearer.

elharo replied:
looks like this could be static since it doesn't depend on instance state (unless I'm missing something)

mbenson commented:
Addressed latest comments... unfortunately I seem to lack permission to re-trigger the CI build.

elharo commented:
checking again https://ci-builds.apache.org/job/Maven/job/maven-box/job/maven-archetype/job/test3/

elharo commented on maven-archetype-plugin/src/main/java/org/apache/maven/archetype/ui/generation/DefaultArchetypeGenerationConfigurator.java:
```diff
@@ -313,46 +308,24 @@ else if ( !archetypeGenerationQueryer.confirmConfiguration( archetypeConfigurati
     {
-        String result = defaultValue;
-        if ( null == result )
+        if ( StringUtils.contains( originalText, "${" ) )
```
could be originalText.contains

In java 7+ we no longer need StringUtils.contains

mbenson replied:
`null` safety

elharo commented on maven-archetype-plugin/src/main/java/org/apache/maven/archetype/ui/generation/DefaultArchetypeGenerationConfigurator.java:
```diff
@@ -313,46 +308,24 @@ else if ( !archetypeGenerationQueryer.confirmConfiguration( archetypeConfigurati
 
-    private String getTransitiveDefaultValue( String defaultValue, ArchetypeConfiguration archetypeConfiguration,
-                                              String requiredProperty, Context context )
+    private static String expandEmbeddedTemplateExpressions( String originalText, String textDescription, Context context )
```
Velocity docs call the textDescription `logTag`. Not sure if that's clearer or not. Up to you.

mbenson replied:
I kind of thought to use these descriptions as a gradual "bridge" over to Velocity's conceptual model.

elharo commented on maven-archetype-plugin/src/main/java/org/apache/maven/archetype/ui/generation/DefaultArchetypeGenerationConfigurator.java:
```diff
@@ -368,71 +341,115 @@ private void restoreCommandLineProperties( ArchetypeConfiguration archetypeConfi
+                final Set<String> referencedPropertyNames = new LinkedHashSet<>();
+
+                String defaultValue = archetypeConfiguration.getDefaultValue( propertyName );
+                if ( StringUtils.contains( defaultValue, "${" ) )
```
defaultValue.contains?

mbenson replied:
`null` safety

elharo commented on maven-archetype-plugin/src/main/java/org/apache/maven/archetype/ui/generation/DefaultArchetypeGenerationConfigurator.java:
```diff
@@ -368,71 +341,115 @@ private void restoreCommandLineProperties( ArchetypeConfiguration archetypeConfi
+        /**
+         * Semantics of this method are "references {@code targetProperty}, {@code sourceProperty} (does)."
+         *
+         * @param targetProperty
```
delete these if you're not going to provide descriptions

mbenson commented:
Thanks @elharo ! What's next?

elharo commented:
I should the final PR in jenkins and merge it if it passs.

elharo commented:
one more time: https://ci-builds.apache.org/job/Maven/job/maven-box/job/maven-archetype/job/test4/

if this passes and there are no more commits, I'll merge

jira-importer commented:
Resolve #393

jira-importer commented:
Resolve #393
